### DISCUSSION TITLE
HiFive Rev B - Getting Started

### DISCUSSION THREAD

**Discussion by kevrenac** [tock/tock#2570] [Q&A] [answered]
Hi,

I've been working on getting started with Tock OS on the HiFive rev B, and after running through the quick-start guide, I have run into a couple roadblocks, was hoping for some help:

- First, I can only load Tock OS via the "make flash-jlink" command. Using "make flash" as called out in the board-specific readme results in errors. First, it was unable to find an FTDI interface. I then updated the .cfg file for the rev B from the config files [here](https://github.com/riscv/riscv-openocd/blob/riscv/tcl/board/sifive-hifive1-revb.cfg), and it then gave me an error of "No J-link device found". It seems to be an openocd error, since even running just the openocd commands give the same issues.
- Since I was able to get things working with the "make flash-jlink" command, I then tried loading apps to it via loading the blink app as an example. That resulted in the error "Size only valid with one TBF". Running "tockloader list" shows no apps installed, as expected. I'll also note that I cannot run just "tockloader [command]", I have to use the "--jlink --board hifive1b" options each time, otherwise tockloader will ask me for the serial port to use, neither of which work (both give the error: Error connecting to bootloader. No "pong" received.)

If anyone has any advice on what to do or where to look to start resolving the issues I've run into, I'd appreciate it. I don't discount the possibility I've done something wrong here, so if you think I've missed a step somewhere, just let me know. Thanks!

**Comment by bradjc** [comment]
Hello!

For the first issue, my best guess is you either need a newer version of openocd or the forked version https://github.com/riscv/riscv-openocd. I unfortunately don't keep too close an eye on openocd/riscv, so I'm not sure what exactly will work.

With tockloader right now yes, you have to tell it what protocol to interface with and which board, otherwise it guesses to use the bootloader serial interface (which the hifive1b doesn't support). There is support for autodetection in the latest commits of tockloader, but not for hifive1b. Although now that I'm thinking about it it shouldn't be hard to add support for hifive1b so you don't need to pass those flags.

As for the error, I can't quite reproduce it. Here is what I did:

```
$ cd tock/boards/hifive1
$ make flash-jlink

$ cd ~/libtock-c/examples/blink
$ git checkout master
$ git pull
$ make RISCV=1
$ tockloader install --board hifive1b --jlink --erase
[INFO   ] No TABs passed to tockloader.
[STATUS ] Searching for TABs in subdirectories.
[WARNING] Warning! There are uncompiled changes!
[WARNING] You may want to run `make` before loading the application.
[INFO   ] Using: ['./build/blink.tab']
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
[STATUS ] Installing app on the board...
[INFO   ] Found sort order:
[INFO   ]   App "blink" at address 0x20040040
[INFO   ] Finished in 23.092 seconds
$ tockloader list --board hifive1b --jlink
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
┌──────────────────────────────────────────────────┐
│ App 0                                            |
└──────────────────────────────────────────────────┘
  Name:                  blink
  Enabled:               True
  Sticky:                False
  Total Size in Flash:   8192 bytes


[INFO   ] Finished in 6.474 seconds
```

Then I get:
```
$ tockloader listen
␛[1F␛[0J␛[?25h␛[0;39;49mBench Clock Reset Complete

ATE0-->ATE0
OK
AT+BLEINIT=0-->OK
AT+CWMODE=0-->ERR CODE:0x01080000
ERROR

HiFive1 initialization complete.
Entering main loop.
Error loading processes!
App memory does not match requested address Actual:0x80002800, Expected:0x80002400
```

So not exactly success, but the app is installed. Possible ideas: I'm running a newer commit of tockloader, it's possible something changed between the version you have and the latest commit. I tried installing the app with tockloader on the v1.6.0 commit and that worked as well, however. Other idea, maybe something is wrong with building the application for riscv? Can you include the output of `make RISCV=1`?

**Reply by bradjc** [reply]
I did get the c_hello app to run by:

```diff
diff --git a/Configuration.mk b/Configuration.mk
index 89580e3..2dc4937 100644
--- a/Configuration.mk
+++ b/Configuration.mk
@@ -56,13 +56,14 @@ TOCK_TARGETS ?= cortex-m0 cortex-m3 cortex-m4 cortex-m7
 else
 # Include the RISC-V targets.
 #  rv32imac|rv32imac.0x20040040.0x80002400 # RISC-V for HiFive1b
 #  rv32imac|rv32imac.0x404*.0x8000*        # RISC-V for arty-e21
 #  rv32imc|rv32imc.0x20030040.0x10003400   # RISC-V for OpenTitan
 TOCK_TARGETS ?= cortex-m0\
                 cortex-m3\
                 cortex-m4\
                 cortex-m7\
-                rv32imac|rv32imac.0x20040040.0x80002400|0x20040040|0x80002400\
+                rv32imac|rv32imac.0x20040040.0x80002800|0x20040040|0x80002800\
                 rv32imac|rv32imac.0x40430060.0x80004000|0x40430060|0x80004000\
                 rv32imac|rv32imac.0x40440060.0x80007000|0x40440060|0x80007000\
                 rv32imc|rv32imc.0x20030080.0x10005000|0x20030080|0x10005000
```

and 

```
$ cd libtock-c/examples/c_hello
$ STACK_SIZE=512 KERNEL_HEAP_SIZE=512 APP_HEAP_SIZE=0 make RISCV=1
$ tockloader install --board hifive1b --jlink --erase
```

The reason the hifive1b doesn't get much attention is it has very little RAM, so it's hard to build any useful apps. The default RAM sizes that libtock-c use are too big for any apps to fit on the hifive1b.

**Comment by kevrenac** [comment]
I'm using the latest openocd from their git repo, but tomorrow I'll try using riscv-openocd to see if that helps, and report back.

I was attempting to install by using the online TAB repo, results are below. If I need to compile/load from the libtock-c git, let me know.
```
[INFO   ] Could not find TAB named "blink" locally.

[0]	No
[1]	Yes

Would you like to check the online TAB repository for that app? [0] 1
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
[STATUS ] Installing app on the board...
[ERROR  ] Size only valid with one TBF
```

**Reply by bradjc** [reply]
Ah, yeah, installing apps that way is currently broken as we switch from tock 1.0 to tock 2.0 (the hosted applications are compiled for a 1.0 kernel). That is on the todo list to fix.

I've opened a couple PRs, one to tock/tock and one tock/libtock-c that fixes two issues I ran into when getting blink to run on the hifive1b. Those should be merged soon.

Also, the issue with openocd might be we may have never actually tried to get it to work for the hifive1b board (only hifive1a). I'm trying with the hifive1b config file, but need to update my openocd version first.

**Comment by bradjc** [comment, marked as answer]
I got tockloader to work with openocd and hifive1b with:

```
tockloader list --board hifive1b --openocd --openocd-board sifive-hifive1-revb.cfg
```

I pushed a commit to tockloader repo that removes the need for the `-openocd-board sifive-hifive1-revb.cfg` flag.

**Reply by kevrenac** [reply]
Thanks! I think my only problem now is with openocd, as when I cloned the libtock-c repo and built `blink` from there, it works perfectly when using J-link:

```
tockloader install blink.tab --jlink --board hifive1b
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
[STATUS ] Installing app on the board...
[INFO   ] Found sort order:
[INFO   ]   App "blink" at address 0x20040040
[INFO   ] Finished in 26.704 seconds

tockloader list --board hifive1b --jlink
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
┌──────────────────────────────────────────────────┐
│ App 0                                            |
└──────────────────────────────────────────────────┘
  Name:                  blink
  Enabled:               True
  Sticky:                False
  Total Size in Flash:   8192 bytes


[INFO   ] Finished in 6.421 seconds
```

But when using openocd, no joy:

```
tockloader list --board hifive1b --openocd --openocd-board sifive-hifive1-revb.cfg
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
[ERROR  ] ERROR: openocd returned with error code 1
[INFO   ] Open On-Chip Debugger 0.11.0+dev-01667-g703960cba (2021-05-04-14:58)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Error: No J-Link device found.


[ERROR  ] openocd error

```
So, I think the only problem now is with openocd. Next step I'm going to try is getting a clean build of openocd, I've got the nightly build, just in case.

**Reply by bradjc** [reply]
I'm using openocd v0.11.0 and that tockloader command is working for me. If you send the output of `tockloader list --board hifive1b --openocd --openocd-board sifive-hifive1-revb.cfg --debug` it may give more clues to what is happening.

Also, why `list` (and other read-only commands seem to work), actually installing apps with openocd isn't working for me. It seems the `fillb` openocd command we use to clear flash gives errors. I opened an issue to see if this is a "just me" problem: https://github.com/riscv/riscv-openocd/issues/602.

**Reply by kevrenac** [reply]
Interesting results. I just compiled the latest from riscv-openocd, still no joy. When I use `--debug` I get the following:

```
tockloader list --board hifive1b --openocd --openocd-board sifive-hifive1-revb.cfg --debug
[INFO   ] Using settings from KNOWN_BOARDS["hifive1b"]
[DEBUG  ] Using read command: "dump_image {{binary}} {address:#x} {length};"
[DEBUG  ] Expanded read command: "dump_image {binary} 0x20040000 200;"
[DEBUG  ] Running "openocd -c " source [find board/sifive-hifive1-revb.cfg]; init; reset init; halt; dump_image /tmp/tmpf0i0_r9z.bin 0x20040000 200;  exit"".
[ERROR  ] ERROR: openocd returned with error code 1
[INFO   ] Open On-Chip Debugger 0.11.0+dev-01667-g703960cba (2021-05-06-15:16)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Error: No J-Link device found.


[ERROR  ] openocd error
```

**Reply by kevrenac** [reply]
Just to confirm something, I'm using a micro-USB cable to connect to the board's on-board debug interface via the J5 connector, not using a JTAG connector (J1) or anything like that. Hopefully that's what I'm supposed to be using.

**Reply by bradjc** [reply]
Yes, I'm just using the normal USB micro connection.

You can run `openocd -c " source [find board/sifive-hifive1-revb.cfg]; init; reset init; halt; dump_image /tmp/tmpf0i0_r9z.bin 0x20040000 200;  exit"` to run the command tockloader is running directly, just to verify nothing strange is happening.

Also, I just plugged in my HiFive1 A and ran `openocd -c " source [find board/sifive-hifive1-revb.cfg]; init; reset init; halt; dump_image /tmp/tmpf0i0_r9z.bin 0x20040000 200;  exit"` and got:

```
$ openocd -c " source [find board/sifive-hifive1-revb.cfg]; init; reset init; halt; dump_image /tmp/tmpf0i0_r9z.bin 0x20040000 200;  exit"
Open On-Chip Debugger 0.11.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Error: No J-Link device found.
```
